gems/sorbet-runtime/test/types/final_method.rb

```diff
@@ -0,0 +1,319 @@
+    Class.new do
+      extend T::Sig
+      def self.foo; end
+      sig(:final) {void}
+      def self.foo; end
+    end
+  end
+
+  it "forbids overriding a final instance method" do
+    c = Class.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final class method" do
+    c = Class.new do
+      extend T::Sig
+      sig(:final) {void}
+      def self.foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method from an included module" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method from an extended module" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend m
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method by including two modules" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m2, m1
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method by extending two modules" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend m2, m1
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "allows calling final methods" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def self.n0; end
+      sig(:final) {params(x: Integer).void}
+      def self.n1(x); end
+      sig(:final) {params(x: Integer, y: Integer).void}
+      def self.n2(x, y); end
+      sig(:final) {params(x: Integer, y: Integer, z: Integer).void}
+      def self.n3(x, y, z); end
+    end
+    m.n0
+    m.n1 1
+    m.n2 1, 2
+    m.n3 1, 2, 3
+  end
+
+  it "calls a user-defined included" do
```

Comment:
Do you have any examples of user-defined `included` and also `final` methods? Could those two features interfere in any way that we'd want to come up with a test case for?

Reply:
Sometimes `included` calls `include` or `extend`, which sounds like something we may want to test.